homeassistant/components/color_extractor/__init__.py

```diff
@@ -0,0 +1,146 @@
+)
+from homeassistant.components.light import (
+    ATTR_RGB_COLOR,
+    DOMAIN as LIGHT_DOMAIN,
+    LIGHT_TURN_ON_SCHEMA,
+    SERVICE_TURN_ON,
+)
+from homeassistant.helpers import aiohttp_client
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+# Extend the existing light.turn_on service schema
+SERVICE_SCHEMA = cv.make_entity_service_schema(
+    {
+        **LIGHT_TURN_ON_SCHEMA,
+        vol.Exclusive(ATTR_PATH, "color_extractor"): cv.isfile,
+        vol.Exclusive(ATTR_URL, "color_extractor"): cv.url,
+    }
+)
+
+
+def _get_file(file_path):
+    """Get a PIL acceptable input file reference.
+
+    Allows us to mock patch during testing to make BytesIO stream.
+    """
+    return file_path
+
+
+async def async_setup(hass, hass_config):
+    """Set up services for color_extractor integration."""
+
+    def _get_color(file_handler) -> tuple:
+        """Given an image file, extract the predominant color from it."""
+        try:
+            color_thief = ColorThief(file_handler)
+        except UnidentifiedImageError as ex:
+            _LOGGER.error("Bad image file provided, are you sure it's an image? %s", ex)
+            return
+
+        # get_color returns a SINGLE RGB value for the given image
+        color = color_thief.get_color(quality=1)
+
+        _LOGGER.debug("Extracted RGB color %s from image", color)
+
+        return color
+
+    async def async_predominant_color_url_service(service_call):
+        """Handle call for URL based image."""
+        service_data = dict(service_call.data)
+
+        url = service_data.pop(ATTR_URL)
+
+        if not hass.config.is_allowed_external_url(url):
+            _LOGGER.error(
+                "External URL '%s' is not allowed, please add to 'allowlist_external_urls'",
+                url,
+            )
+            return
+
+        _LOGGER.debug("Getting predominant RGB from image URL '%s'", url)
+
+        # Download the image into a buffer for ColorThief to check against
+        try:
+            session = aiohttp_client.async_get_clientsession(hass)
+
+            with async_timeout.timeout(10):
+                response = await session.get(url)
+
+        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
+            _LOGGER.error("Failed to get ColorThief image due to HTTPError: %s", err)
+            return
+
+        content = await response.content.read()
+
+        with io.BytesIO(content) as _file:
+            _file.name = "color_extractor.jpg"
+            _file.seek(0)
+
+            color = _get_color(_file)
+
+        if color:
+            service_data[ATTR_RGB_COLOR] = color
+
+            await hass.services.async_call(
+                LIGHT_DOMAIN, SERVICE_TURN_ON, service_data, blocking=True
+            )
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_PREDOMINANT_COLOR_URL,
+        async_predominant_color_url_service,
+        schema=SERVICE_SCHEMA,
```

Comment:
This is not correct. You are now allowing to pass in a path to this service as well.

Reply:
You should have 1 service handler that handles both cases, that way your service is easier to reason about.

Reply:
Sorry, I'm not sure I fully understand. I thought adding the Exclusive with the group named `color_extractor`, that either `color_extract_path` or `color_extract_url` had to be provided, is that not the case?

As that would allow sharing the same schema right?

Thanks for all your feedback btw, this is going to be pretty useful once all finished :)

Reply:
I think I've understood this. I now have one service `color_extractor.turn_on`, if `color_extract_url` is present in the service data we call the url handler, and if `color_extract_path` is provided, it goes to the file handler.